### DISCUSSION TITLE
useRouter return null in ReactDOM.render

### DISCUSSION THREAD

**Discussion by javaxiu** [vercel/next.js#37908] [Help]
in this case:
```typescript

const ConfirmDialog = () => {
     const router = useRouter(); // router got null here
     return <div></div>
}
const confirm = () => {
   ReactDOM.render(
      <ConfirmDialog />,
      container,
    );
}

const AnyPage = () => {
    const show = () => {
      confirm();
   }
   return <div>a normal page here</div>
}
```

useRouter got null if the context component was render by `ReactDOM.render`
I've try to search, it seems someone has met this before ?  https://www.reddit.com/r/nextjs/comments/s9hm7z/userouter_null_when_used_with_reactdomrender/

**Comment by icyJoseph** [comment]
When you invoke `ReactDOM.render(<App/>)` - you are creating a separate React tree, and mounting it on `container`. This tree has no access to Context Providers on the Next.js App tree, so that's why `useRouter` returns `null`, it can't find its context provider.

Use Portals for this, https://reactjs.org/docs/portals.html, it achieves the same, and it doesn't come at the cost given above. There's also a bunch of reasons to no use `ReactDOM.render` like that, but yours is a clear example of the consequences.

Portals do exactly what you want here, and even more, https://medium.com/hackernoon/using-a-react-16-portal-to-do-something-cool-2a2d627b0202.
